src/api/python/speechd/client.py

```diff
@@ -745,14 +823,20 @@ def speak(self, text, callback=None, event_types=None):
         message is queued on the server and the method returns immediately.
 
         """
-        self._conn.send_command('SPEAK')
-        result = self._conn.send_data(text)
         if callback:
-            msg_id = int(result[2][0])
-            # TODO: Here we risk, that the callback arrives earlier, than we
-            # add the item to `self._callback_handler'.  Such a situation will
-            # lead to the callback being ignored.
-            self._callback_handler.add_callback(msg_id, callback, event_types)
+            self._callback_handler.begin_add_callback()
+        try:
+            self._conn.send_command('SPEAK')
+            result = self._conn.send_data(text)
+            if callback:
+                msg_id = int(result[2][0])
+        except:
+            if callback:
+                self._callback_handler.cancel_add_callback()
+            raise
+        else:
+            if callback:
+                self._callback_handler.end_add_callback(msg_id, callback, event_types)
```

Comment:
This is a bit convoluted and I would have loved providing a simpler API, maybe like so (100% untested):
```python

    def add_callback(self, callback=None, event_types=None):
        """
        with handler.add_callback(callback, event_types) as request:
            connection.send_command('SPEAK')
            result = connection.send_data(data)
            request.set_message_id(int(result[2][0]))
        """
        class AddCallbackRequest:
            def __init__(self, handler, callback=None, event_types=None):
                self._handler = handler
                self._msg_id = None
                self._callback = callback
                self._event_types = event_types

            def set_message_id(self, msg_id):
                self._msg_id = msg_id

            def __enter__(self):
                self.handler.begin_add_callback()
                return self

            def __exit__(self, exc_type, exc_value, traceback):
                if exc_type:
                    self.handler.cancel_add_callback()
                else:
                    assert self._msg_id is not None
                    self.handler.end_add_callback(self._msg_id, self._callback, self._event_types)
                return False

        return AddCallbackRequest(self, callback, event_types)
```

but unfortunately as this (sole) user has a conditional `callback` it's not really usable. Meh.

Reply:
I would be fine with something like this:

```
        if callback:
            with handler.add_callback(callback, event_types) as request:
                self._conn.send_command('SPEAK')
                result = self._conn.send_data(text)
                request.set_message_id(int(result[2][0]))
        else:
            self._conn.send_command('SPEAK')
            result = self._conn.send_data(text)
```